druid/src/text/rich_text.rs

```diff
@@ -0,0 +1,193 @@
+};
+use crate::{ArcStr, Data, Env, FontDescriptor, KeyOrValue};
+
+/// Text with optional style spans.
+#[derive(Debug, Clone, Data)]
+pub struct RichText {
+    buffer: ArcStr,
+    attrs: Arc<AttributeSpans>,
+}
+
+impl RichText {
+    /// Create a new `RichText` object with the provided text.
+    pub fn new(buffer: ArcStr) -> Self {
+        RichText::new_with_attributes(buffer, Default::default())
+    }
+
+    /// Create a new `RichText`, providing explicit attributes.
+    pub fn new_with_attributes(buffer: ArcStr, attributes: AttributeSpans) -> Self {
+        RichText {
+            buffer,
+            attrs: Arc::new(attributes),
+        }
+    }
+
+    /// Builder-style method for adding an [`Attribute`] to a range of text.
+    ///
+    /// [`Attribute`]: enum.Attribute.html
+    pub fn with_attribute(mut self, range: impl RangeBounds<usize>, attr: Attribute) -> Self {
+        self.add_attribute(range, attr);
+        self
+    }
+
+    /// The length of the buffer, in utf8 code units.
+    pub fn len(&self) -> usize {
+        self.buffer.len()
+    }
+
+    /// Returns `true` if the underlying buffer is empty.
+    pub fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    /// Add an [`Attribute`] to the provided range of text.
+    ///
+    /// [`Attribute`]: enum.Attribute.html
+    pub fn add_attribute(&mut self, range: impl RangeBounds<usize>, attr: Attribute) {
+        let range = util::resolve_range(range, self.buffer.len());
+        Arc::make_mut(&mut self.attrs).add(range, attr);
+    }
+}
+
+impl PietTextStorage for RichText {
+    fn as_str(&self) -> &str {
+        self.buffer.as_str()
+    }
+}
+
+impl TextStorage for RichText {
+    fn add_attributes(
+        &self,
+        mut builder: PietTextLayoutBuilder,
+        env: &Env,
+    ) -> PietTextLayoutBuilder {
+        for (range, attr) in self.attrs.to_piet_attrs(env) {
+            builder = builder.range_attribute(range, attr);
+        }
+        builder
+    }
+}
+
+/// A builder for [`RichText`].
+///
+/// # Example
+/// ```
+/// # use druid::text::{RichTextBuilder, Attribute};
+/// let mut rich_text = RichTextBuilder::new();
+/// rich_text.push("Hello World").underline(true);
+/// let rich_text = rich_text.build();
+/// ```
+///
+/// [`RichText`]: RichText
+#[derive(Debug, Default)]
+pub struct RichTextBuilder {
+    buffer: String,
+    attrs: AttributeSpans,
+}
+
+impl RichTextBuilder {
+    /// Create a new `RichTextBuilder`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add string to the end of text and `AttributesAdder` for the added string.
+    pub fn push(&mut self, string: &str) -> AttributesAdder {
+        let range = self.buffer.len()..(self.buffer.len() + string.len());
+        self.buffer.push_str(string);
+        self.range(range)
+    }
+
+    /// Get the `AttributesAdder` for the range.
+    pub fn range(&mut self, range: Range<usize>) -> AttributesAdder {
```

Comment:
I'd give this a slightly more verbose name:

```suggestion
    /// Get an [`AttributesAdder`] for the given range.
    ///
    /// This can be used to modify styles for a given range after it has been added.
    pub fn add_attributes_for_range(&mut self, range: Range<usize>) -> AttributesAdder {
```